cluster/cluster.go

```diff
@@ -15,4 +38,85 @@ func Init(name, version string, primary bool, started time.Time) {
 		primaryChange: time.Now(),
 		stateChange:   time.Now(),
 	}
+	Mode = m
+	go probePeers(interval)
+}
+
+func Stop() {
+	close(shutdown)
+}
+
+func AddPeer(remoteAddr *url.URL) {
+	mu.Lock()
+	peer := &Node{
+		remoteAddr:  remoteAddr,
+		stateChange: time.Now(),
+	}
+	peers = append(peers, peer)
+	go peer.Probe()
+	mu.Unlock()
+}
+
+func GetPeers() []*Node {
+	mu.Lock()
+	p := make([]*Node, len(peers))
+	copy(p, peers)
+	mu.Unlock()
+	return p
+}
+
+func probePeers(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for {
+		select {
+		case <-shutdown:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			mu.Lock()
+			for _, peer := range peers {
+				go peer.Probe()
+			}
+			mu.Unlock()
+		}
+	}
+}
+
+// return the list of peers to broadcast requests to
+// Only 1 peer per partition is returned
+func PeersForQuery() []*Node {
+	// If we are running in single mode, just return thisNode
+	if Mode == ModeSingle {
+		return []*Node{ThisNode}
+	}
+
+	peersMap := make(map[int32][]*Node)
+	for _, part := range ThisNode.GetPartitions() {
+		peersMap[part] = []*Node{ThisNode}
+	}
+	mu.Lock()
+	for _, peer := range peers {
+		if !peer.IsReady() {
+			continue
+		}
+		for _, part := range peer.GetPartitions() {
+			peersMap[part] = append(peersMap[part], peer)
+		}
+	}
+	mu.Unlock()
+	selectedPeers := make(map[string]*Node)
+	answer := make([]*Node, 0)
+	// we only need to query 1 node for each partition.
+	// we are just using the first node for each partition, which will be
+	// ThisNode if ThisNode owns the partition.  Later we can add more
+	// complex scheduling logic to distribute queries across the cluster
```

Comment:
yes this is simple, but gets the job done.  It is by no means the best solution, but i think it is good enough for now.

Assume we have 4 partitions and 4 nodes

nodeA | nodeB | nodeC | nodeD
--------- | -------- | ---------- | -------
0,1 | 0, 1 | 2, 3 | 2, 3

if  a query is sent to nodeA, it will select nodeA for partitions 0,1 and either nodeC or nodeD for partitions 2,3.    So every query will only map to 2 nodes, the local node and 1 other.

The only problem that i see is that both nodeA and nodeB could consistently send queries to nodeC and never nodeD.